### ISSUE TITLE
License Key: Make configurable via wp-config.php

### ISSUE THREAD

**Issue by JanThiel** [AffiliateWP/AffiliateWP#1229]
Hi,

when running AffiliateWP in a Multisite, you need to set the License Key on each site manually. If you want the updates to work, it also has to be activated on the Networks main page. Although if you don't wanna use it there. Further more when using automatic deployments, each UI input for license keys is breaking scalability.

Simple Solution: Offer an alternative way to set the license key in wp-config.php instead of UI + this->options.
So check for is_defined("wpaff_license") before checking the this->settings->license_key and continue stuff with that value instead.

In the Admin Backend the key should not be visible or editable. Just to make clear it is set at another place and not there. As it cannot and should not be edited in the WP-Admin backend.

Thanks, Best Regards,

Jan

**Comment by pippinsplugins** [comment]
Yes! Let's do it.

**Comment by DrewAPicture** [comment]
Added a PR: #1268

**Comment by JanThiel** [comment]
@DrewAPicture There are two more things I would like to throw into for consideration. I also did some implementation work on this one (your PR made my code obsolete, which is fine anyway) so I would like to share my thoughts:

1. The License Key Section in the wp-admin backend should be disabled when the code is set via define( ... ). I implemented a "disabled='disabled'" for the input field, as well as displaying "*********..." instead of the real license. Just to make it clear that the code is set otherwere. The buttons to disable and activate a license should be obsolete as well. That is, because you cannot actually achieve anything with the input field, when a value is defined in wp-config (which is good!). So the UI should reflect that.

2. If the key is set in wp-config, you need to trigger the "Check license" initially in the code. Otherwise the license will never be activated. That is the point I haven't solved neither. But you might know the appropriate piece of code where it belongs to. Check should be like if license != active and defined(...) try to activate it.

That's from my side, thanks for the quick implementation :-)

Jan

**Comment by pippinsplugins** [comment]
I'm not sure clicking `Activate` from the UI makes sense when the license key is entered in the wp-config.php file but I'm also not sure auto-activating it makes sense either. The main scenario I'm thinking of is when a license key has been entered in the wp-config.php file but is invalid for one reason or another. In that case, we'll still need to show some sort of feedback to the user. Currently that feedback is shown after clicking `Activate` or after the `check_license()` action has run.

We could attempt to the activate the license during the install routine if it is set in wp-config.php and then just fallback to the existing `check_license()` method for periodically validating it.

**Comment by JanThiel** [comment]
You do have the checks for the Admin Notice of an invalid license. That was the spot where I was thinking about to perform the automatic activation. That should also be a good way to display an invalid license which was entered into wp-config but failed via auto activation. As you know about a license entered in wp-config and can change your handling and the error texts appropriately.
Might that be a proper solution?

Install routine (or better "Activating the Plugin" routine) might also be a good idea as a "fallback" but should not be necessary when you use the generic approach as written above.

At all, the License Area in the wp-admin config should be completely disabled or hidden (which I wouldn't do to keep the UI experience the same for both ways) to not lure people into a wrong expectation.

Brgds,

Jan

**Comment by pippinsplugins** [comment]
@DrewAPicture this seems to work perfectly, with one caveat: while it is possible to deactivate a license, it automatically reactivates on the next page load.

I see two options for it:

1. Do not show `Deactivate` license if globally defined
2. Store a special status if the license key is deactivated when globally defined and only auto activate if this status is not set.

**Comment by DrewAPicture** [comment]
Yeah, it's basically in an incomplete state, so both of those points (and
some other small things) need to be addressed yet.

**Comment by pippinsplugins** [comment]
Ah ok, otherwise it worked great!

**Comment by DrewAPicture** [comment]
@pippinsplugins OK, give this another test.

Per yesterday's [feedback](https://github.com/AffiliateWP/AffiliateWP/issues/1229#issuecomment-228934910), I've made activating and deactivating a conscious decision. The field will always be populated with the global key if it's defined, but imo the user should be able to activate or deactivate whether it's defined or not.

**Comment by pippinsplugins** [comment]
Works great! Just one comment about the pre registered settings action (see PR).

**Comment by DrewAPicture** [comment]
@JanThiel This will ship with the 1.9 release.

Essentially what we did was two-fold:

1. Made activation and deactivation a conscious choice, meaning that you can globally supply the license key via the constant, but you still have to click Activate or Deactivate. Auto-activation can be made to be intuitive, but generally it's annoying.

![Global license](http://cl.ly/0o2r3R0c2i1K/Screen%20Shot%202016-06-30%20at%201.16.56%20PM.png)

2. If a global license key is defined, the license key field still displays, but populates with the value of the constant, is disabled, and a special description is added explaining why it's not editable from that screen.

**Comment by JanThiel** [comment]
Hey @DrewAPicture, thanks for the work! And sorry for getting back so lately to you with any kind of reply... Still in WordCamp Europe aftermath mode...

As this is solving the general configurability via wp-config it does not solve the scalability aspect I initially had in mind (see in my initial post). So again there is UI interaction necessary to enable a license. That doesn't work for automatic deployments.

So picking up at the point where you require the conscious "Activation" or "Deactivation" of the license - I totaly understand that this will be necessary and useful for the vast majority of the users. To still achieve the scalability and automation, why not add another option that completely removes the license key section from the backend and does activation automatically. Deactivating a license can still be achieved via affiliatewp.com backend. Or it doesn't matter because - as we do - use an unlimited license to actually and totaly get rid of the hassle.

So, any chance of adding this to allow completely automated deployments?

Thanks alot so far guys, best regards,

Jan

**Comment by pippinsplugins** [comment]
We can definitely still consider adding in automatic activation but I think we should go ahead and split that off into a new issue where we can discuss the implementation.

**Comment by JanThiel** [comment]
Will do that Pippin. Already thought about that...

**Comment by pippinsplugins** [comment]
Note, new constant is `AFFILIATEWP_LICENSE_KEY`.
